File: MyApplication/scripts/lib/workspace-paths.mjs
/**
 * 宿主运行资料的唯一落盘策略。工程源码与 SDK 原生输出不经过本模块；自建日志、
 * 测试夹具、Native 构建树必须在工作区 .workspace 或仓库外的系统临时目录。
 * 解析只返回路径，不创建目录，也不修改全局环境变量或删除旧资料。
 */
import { existsSync, mkdirSync, mkdtempSync, realpathSync } from 'node:fs';
import { createHash } from 'node:crypto';
import os from 'node:os';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

export const PROJECT_ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '../..');
const RUN_STAMP = `${new Date().toISOString().replace(/[-:]/g, '').slice(0, 15)}-${process.pid}`;

/** 解析最近存在的祖先，防止 junction/symlink 把看似外置的路径重新导回源码仓库。 */
function physicalPath(value) {
  let ancestor = path.resolve(value);
  const suffix = [];
  while (!existsSync(ancestor)) {
    const parent = path.dirname(ancestor);
    if (parent === ancestor) break;
    suffix.unshift(path.basename(ancestor));
    ancestor = parent;
  }
  return path.resolve(existsSync(ancestor) ? realpathSync(ancestor) : ancestor, ...suffix);
}

function inside(parent, candidate) {
  const relative = path.relative(parent, candidate);
  return relative === '' || (!relative.startsWith(`..${path.sep}`) && relative !== '..' && !path.isAbsolute(relative));
}

/** 显式路径也必须外置；既保护当前工程，也保护公开快照外层仓库及其他现有 Git 项目。 */
export function assertExternalOutputPath(value, { root = PROJECT_ROOT } = {}) {
  const candidate = physicalPath(value);
  const project = physicalPath(root);
  if (inside(project, candidate)) throw new Error(`Output must be outside the project: ${value}`);
  for (let current = candidate; ; current = path.dirname(current)) {
    if (existsSync(path.join(current, '.git'))) {
      throw new Error(`Output must be outside Git repositories: ${value}`);
    }
    if (path.dirname(current) === current) break;
  }
  return path.resolve(value);
}

/** 使用显式工作区优先；只有同时有 .workspace 与根规则/规划标记的祖先才自动认作工作区。 */
function workspaceRoot(root, env) {
  if (env.AMCL_WORKSPACE_ROOT) {
    if (!path.isAbsolute(env.AMCL_WORKSPACE_ROOT)) throw new Error('AMCL_WORKSPACE_ROOT must be absolute');
    return path.resolve(env.AMCL_WORKSPACE_ROOT);
  }
  for (let current = path.dirname(root); ; current = path.dirname(current)) {
    if (existsSync(path.join(current, '.workspace')) &&
        (existsSync(path.join(current, 'AGENTS.md')) || existsSync(path.join(current, 'WORKSPACE-ORGANIZATION-PLAN.md')))) {
      return current;
    }
    if (path.dirname(current) === current) return null;
  }
}

function checkedId(id) {
  if (!/^[a-z0-9][a-z0-9-]{0,47}$/i.test(id) || /^(con|prn|aux|nul|com[1-9]|lpt[1-9])$/i.test(id)) {
    throw new Error(`Invalid short workspace id: ${id}`);
  }
  return id;
}

/**
 * kind=build/wt 的目录按工程真实绝对路径哈希隔离，便于跨进程复用缓存且不混用不同克隆。
 * kind=run 默认每次进程独立；AMCL_RUN_DIR 可将多条命令归入同一显式任务目录。
 * 独立克隆/CI 没有本机工作区时使用系统临时容器；系统 TMP 若被设到仓内会明确失败。
 */
export function resolveWorkspacePath(kind, id, { root = PROJECT_ROOT, env = process.env } = {}) {
  if (!['run', 'build', 'wt'].includes(kind)) throw new Error(`Invalid workspace kind: ${kind}`);
  checkedId(id);
  root = physicalPath(root);
  const identity = process.platform === 'win32' ? root.toLowerCase() : root;
  const checkout = createHash('sha256').update(identity).digest('hex').slice(0, 8);
  const workspace = workspaceRoot(root, env);
  const container = workspace ? path.join(workspace, '.workspace') : path.join(os.tmpdir(), 'amcl-workspace');
  let output;
  if (kind === 'run' && env.AMCL_RUN_DIR) {
    if (!path.isAbsolute(env.AMCL_RUN_DIR)) throw new Error('AMCL_RUN_DIR must be absolute');
    output = path.join(env.AMCL_RUN_DIR, id);
  } else {
    output = path.join(container, kind, `${id}-${checkout}${kind === 'run' ? `-${RUN_STAMP}` : ''}`);
  }
  return assertExternalOutputPath(output, { root });
}

/** 各调用者只追加受控相对文件名；不允许通过绝对路径或 .. 绕过上述边界。 */
export function workspacePath(kind, id, ...parts) {
  for (const part of parts) {
    if (path.isAbsolute(part) || part.split(/[\\/]/).includes('..')) throw new Error(`Invalid output suffix: ${part}`);
  }
  return assertExternalOutputPath(path.join(resolveWorkspacePath(kind, id), ...parts));
}

/** 返回宿主测试临时目录的父目录。清理由创建该临时子目录的测试负责，不能删除整个共享父目录。 */
export function workspaceTempRoot() {
  const directory = workspacePath('build', 'host-tests');
  mkdirSync(directory, { recursive: true });
  return directory;
}

/** 独占目录防止并行测试覆盖；前缀仍由调用者决定，便于对应已有清理边界检查。 */
export function createWorkspaceTemp(prefix, { kind = 'build', root = PROJECT_ROOT } = {}) {
  if (!prefix || /[\\/]/.test(prefix) || prefix === '..') throw new Error('Invalid temporary directory prefix');
  const directory = resolveWorkspacePath(kind, 'host-tests', { root });
  mkdirSync(directory, { recursive: true });
  return mkdtempSync(path.join(directory, prefix));
}

// PowerShell 通过此 CLI 复用策略；Python/WSL 使用等价的独立实现，并由跨语言测试防止漂移。
// 参数是独立 argv，避免 shell 拼接。
if (process.argv[1] && path.resolve(process.argv[1]) === fileURLToPath(import.meta.url)) {
  try {
    const [kind, id, root = PROJECT_ROOT, explicit] = process.argv.slice(2);
    console.log(explicit ? assertExternalOutputPath(explicit, { root }) : resolveWorkspacePath(kind, id, { root }));
  } catch (error) {
    console.error(error.message);
    process.exitCode = 1;
  }
}

File: MyApplication/scripts/lib/workspace_paths.py
"""Python/WSL 宿主工具的路径策略；与 workspace-paths.mjs 等价，不要求 Linux 安装 Node。"""
from pathlib import Path
from datetime import datetime, timezone
import hashlib
import os
import re
import tempfile

PROJECT_ROOT = Path(__file__).resolve().parents[2]
RUN_STAMP = datetime.now(timezone.utc).strftime("%Y%m%dT%H%M%S") + "-" + str(os.getpid())


def external_output_path(value, root=None):
    """解析链接后核验仓库边界；公开仓内的嵌套 MyApplication 也不得把输出放到外层仓库。"""
    candidate, project = Path(value).resolve(), Path(root or PROJECT_ROOT).resolve()
    if candidate == project or project in candidate.parents:
        raise ValueError("Output must be outside the project: " + str(value))
    if any((parent / ".git").exists() for parent in [candidate, *candidate.parents]):
        raise ValueError("Output must be outside Git repositories: " + str(value))
    return candidate


def workspace_path(kind, task_id, *parts):
    """返回外部输出位置，不创建目录；配置优先、标记发现其次，独立克隆退回仓外系统临时目录。"""
    if kind not in ("run", "build", "wt"):
        raise ValueError("Invalid workspace kind: " + kind)
    if not re.fullmatch(r"[a-zA-Z0-9][a-zA-Z0-9-]{0,47}", task_id) or re.fullmatch(r"con|prn|aux|nul|com[1-9]|lpt[1-9]", task_id, re.I):
        raise ValueError("Invalid short workspace id: " + task_id)
    for part in parts:
        if Path(part).is_absolute() or ".." in Path(part).parts:
            raise ValueError("输出子路径不得越过任务目录")
    workspace = os.environ.get("AMCL_WORKSPACE_ROOT")
    if workspace:
        if not Path(workspace).is_absolute():
            raise ValueError("AMCL_WORKSPACE_ROOT must be absolute")
        workspace = Path(workspace)
    else:
        workspace = next((parent for parent in PROJECT_ROOT.parents if (parent / ".workspace").exists()
            and ((parent / "AGENTS.md").exists() or (parent / "WORKSPACE-ORGANIZATION-PLAN.md").exists())), None)
    identity = str(PROJECT_ROOT).lower() if os.name == "nt" else str(PROJECT_ROOT)
    checkout = hashlib.sha256(identity.encode("utf-8")).hexdigest()[:8]
    container = workspace / ".workspace" if workspace else Path(tempfile.gettempdir()) / "amcl-workspace"
    run = os.environ.get("AMCL_RUN_DIR") if kind == "run" else None
    if run:
        if not Path(run).is_absolute():
            raise ValueError("AMCL_RUN_DIR must be absolute")
        result = Path(run) / task_id
    else:
        result = container / kind / (task_id + "-" + checkout + ("-" + RUN_STAMP if kind == "run" else ""))
    return external_output_path(result.joinpath(*parts))


def temporary_directory(*args, **kwargs):
    """临时测试保留 tempfile 的自动回收语义，但将默认父目录收口到外部宿主测试区。"""
    if "dir" not in kwargs:
        directory = workspace_path("build", "host-tests")
        directory.mkdir(parents=True, exist_ok=True)
        kwargs["dir"] = directory
    else:
        # 显式 tempfile 父目录也不能绕过外置规则；只有原子文件替换可使用目标同级暂存文件。
        kwargs["dir"] = external_output_path(kwargs["dir"])
    return tempfile.TemporaryDirectory(*args, **kwargs)


if __name__ == "__main__":
    # Bash/无 Node 的 Linux 测试镜像共用此入口；只解析路径，mkdir/生命周期仍由调用者管理。
    import argparse
    parser = argparse.ArgumentParser(description="Resolve AMCL host output outside source repositories")
    parser.add_argument("kind", choices=("run", "build", "wt"))
    parser.add_argument("task_id")
    parser.add_argument("--explicit", default="")
    arguments = parser.parse_args()
    try:
        print(external_output_path(arguments.explicit) if arguments.explicit else workspace_path(arguments.kind, arguments.task_id))
    except ValueError as error:
        parser.exit(1, str(error) + "\n")
